lib/ansible/modules/database/mysql/mysql_db.py

```diff
@@ -346,81 +373,65 @@ def main():
+            existence_list.append(each_database)
+        else:
+            non_existence_list.append(each_database)
+
+    if state == "absent":
+        if module.check_mode:
+            module.exit_json(changed=bool(existence_list), db=db_name, db_list=db)
+        try:
+            changed = db_delete(cursor, existence_list)
+        except Exception as e:
+            module.fail_json(msg="error deleting database: %s" % to_native(e))
+        module.exit_json(changed=changed, db=db_name, db_list=db)
+    elif state == "present":
+        if module.check_mode:
+            module.exit_json(changed=bool(non_existence_list), db=db_name, db_list=db)
+        changed = False
+        if non_existence_list:
             try:
-                changed = db_delete(cursor, db)
+                changed = db_create(cursor, non_existence_list, encoding, collation)
             except Exception as e:
-                module.fail_json(msg="error deleting database: %s" % to_native(e))
-            module.exit_json(changed=changed, db=db_name)
-
-        elif state == "dump":
-            if module.check_mode:
-                module.exit_json(changed=True, db=db_name)
-            rc, stdout, stderr = db_dump(module, login_host, login_user,
-                                         login_password, db, target, all_databases,
-                                         login_port, config_file, socket, ssl_cert, ssl_key,
-                                         ssl_ca, single_transaction, quick, ignore_tables)
-            if rc != 0:
-                module.fail_json(msg="%s" % stderr)
-            else:
-                module.exit_json(changed=True, db=db_name, msg=stdout)
-
-        elif state == "import":
-            if module.check_mode:
-                module.exit_json(changed=True, db=db_name)
+                module.fail_json(msg="error creating database: %s" % to_native(e),
+                                 exception=traceback.format_exc())
+        module.exit_json(changed=changed, db=db_name, db_list=db)
+    elif state == "dump":
+        if module.check_mode:
+            module.exit_json(changed=not bool(non_existence_list), db=db_name, db_list=db)
+        if non_existence_list:
+            module.fail_json(msg="Cannot dump database(s) %r - not found" % (', '.join(non_existence_list)))
+        rc, stdout, stderr = db_dump(module, login_host, login_user,
+                                     login_password, db, target, all_databases,
+                                     login_port, config_file, socket, ssl_cert, ssl_key,
+                                     ssl_ca, single_transaction, quick, ignore_tables)
+        if rc != 0:
+            module.fail_json(msg="%s" % stderr)
+        module.exit_json(changed=True, db=db_name, db_list=db, msg=stdout)
+    elif state == "import":
+        changed = True
+        if module.check_mode:
+            module.exit_json(changed=True, db=db_name, db_list=db)
```

Comment:
Let's consider a situation where all the databases to be imported already exists. In this case, the changed state for check mode would be `False` since `changed=bool(non_existence_list)`. This contradicts to the previous `mysql_db` code for import where it always returns `True` irrespective of whether a database exists or not.

Previous code for `state=import`:
- db exists [code](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/mysql/mysql_db.py#L372)
- db not exist [code](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/mysql/mysql_db.py#L400)

Reply:
You're right, it always returned `True` before. There's a difference to before, though: before, `db_create` *always* returned `True`. Now, it only returns `True` if creating all DBs succeeded. So the `if changed:` can be removed in line 426, and the `if` block can always be executed. As long as `db_create` didn't result in an exception (in which case the module dies), all DBs should be there anyway.

Reply:
Agree that `if changed:` could be removed and `if` block can always be executed however, I still think that `changed` should be equals to `True` instead of `changed=bool(non_existence_list)` since import functionality also overrides imported database if that already exists.
This will again lead to system change, right?

Reply:
Discussed the same with @felixfontein and we agree that `changed` state would be always `True`